File: MP4/sources/student/Python/proctest.py
#!/usr/bin/env python

# --------------------------------------------------------------------------- #
# Developer: Andrew Kirfman                                                   #
# Project: CSCE-313 Machine Problem #4                                        #
#                                                                             #
# File: ./student/proctest.py                                                 #
# --------------------------------------------------------------------------- #


# --------------------------------------------------------------------------- #
# Standard Library Includes                                                   #
# --------------------------------------------------------------------------- #

import sys
import re
from subprocess import check_output, CalledProcessError


"""
Note to students: Your program must be written using the provided class
template below.  Functions that require your implementation are marked
with "Work needed here" in their method docstrings.

Be sure to delete all of these extraneous comments before submitting your
code for grading.  Also, delete all extra 'pass' statements.  They are
noop instructions and only there because Python does not like empty functions.
"""


class Proctest(object):
    """
    Object to contain process data and getter functions
    """

    def __init__(self, pid):
        """
        Read all data from pertinent files and save as class members

        Work needed here!
        """
        pass


    def getpid(self):
        """
        Returns the process id

        Work needed here!
        """
        pass


    def getppid(self):
        """
        Returns process's parent's ID

        Work needed here!
        """
        pass


    def geteuid(self):
        """
        Returns the process's effective user ID

        Work needed here!
        """
        pass


    def getegid(self):
        """
        Returns process's effective group ID

        Work needed here!
        """
        pass


    def getruid(self):
        """
        Returns process's real user ID

        Work needed here!
        """
        pass


    def getrgid(self):
        """
        Returns process's real group ID

        Work needed here!
        """
        pass


    def getfsuid(self):
        """
        Returns process's file system user ID

        Work needed here!
        """
        pass


    def getfsgid(self):
        """
        Returns process's file system group ID

        Work needed here!
        """
        pass


    def getstate(self):
        """
        Returns process's current state

        Work needed here!
        """
        pass


    def getthread_count(self):
        """
        Returns process's current thread count

        Work needed here!
        """
        pass


    def getpriority(self):
        """
        Returns process's current prioroty number

        Work needed here!
        """
        pass


    def getniceness(self):
        """
        Returns process's current niceness value

        Work needed here!
        """
        pass


    def getstime(self):
        """
        Returns process's current stime value

        Work needed here!
        """
        pass


    def getutime(self):
        """
        Returns process's current utime value

        Work needed here!
        """
        pass


    def getcstime(self):
        """
        Returns process's current cstime value

        Work needed here!
        """
        pass


    def getcutime(self):
        """
        Returns process's current cutime value

        Work needed here!
        """
        pass


    def getstartcode(self):
        """
        Returns process's startcode

        Work needed here!
        """
        pass


    def getendcode(self):
        """
        Returns process's endcode

        Work needed here!
        """
        pass


    def getesp(self):
        """
        Returns process's stack pointer

        Work needed here!
        """
        pass


    def geteip(self):
        """
        Returns process's instruction pointer

        Work needed here!
        """
        pass


    def getfiles(self):
        """
        Returns process's current number of open file descriptors

        Work needed here!
        """
        pass


    def getvoluntary_context_switches(self):
        """
        Returns the number of times that the process has voluntarily
        yielded control to the kernel.

        Work needed here!
        """
        pass


    def getnonvoluntary_context_switches(self):
        """
        Returns the number of times that the process has had control
        taken from it forcefully by the kernel.

        Work needed here!
        """
        pass


    def getlast_cpu(self):
        """
        Returns the last cpu that the process executed on

        Work needed here!
        """
        pass


    def getallowed_cpus(self):
        """
        Returns a list of allowed processors

        Work needed here!
        """
        pass


    def getmemory_map(self):
        """
        Returns process's memory map

        Work needed here!
        """
        pass


def main():
    """
    See comments below for expected content of this function

    Work needed here!
    """

    # Read in PID

    # Instantiate class member

    # Print out values

    pass


if __name__ == "__main__":
    main()

File: MP4/sources/student/C++/proctest.cpp
/* ------------------------------------------------------------------------- */
/* Developer: Andrew Kirfman                                                 */
/* Project: CSCE-313 Machine Problem #4                                      */
/*                                                                           */
/* File: ./proctest.cpp (student version)                                    */
/* ------------------------------------------------------------------------- */


/* ------------------------------------------------------------------------- */
/* Standard Library Includes                                                 */
/* ------------------------------------------------------------------------- */

#include<iostream>
#include<vector>
#include<stdlib.h>
#include<stdio.h>
#include<unistd.h>
#include<fcntl.h>
#include<string.h>
#include<string>
#include<regex>
#include<sys/types.h>
#include<sys/stat.h>
#include<sys/wait.h>

/* ------------------------------------------------------------------------- */
/* User Defined Includes                                                     */
/* ------------------------------------------------------------------------- */

#include "proctest.h"

/* ------------------------------------------------------------------------- */
/* Proctest Private Member Functions                                         */
/* ------------------------------------------------------------------------- */

/* 
 * This function turns a char* buffer into a string given the address
 * to the buffer.  
 */ 
std::string Proctest::stringify_buffer(char *buffer)
{
    std::string buffer_string = "";
    int pos = 0;

    while(true)
    {
        if(buffer[pos] == '\0')
        {
            break;
        }
        else
        {
            buffer_string += buffer[pos];
            pos++;
        }
    }

    return buffer_string;
}

/* 
 * This function takes in a string and breaks it up into a vector
 * of space separated words.  
 */ 
std::vector<std::string> Proctest::split_on_spaces(std::string input)
{
    std::vector<std::string> components;
    std::string token = "";

    for(unsigned short int i=0; i<input.length(); i++)
    {
        if(input[i] != ' ')
        {
            token += input[i];
        }
        else
        {
            components.push_back(token);
            token = "";
        }
    }

    return components;
}

/* ------------------------------------------------------------------------- */
/* Proctest Public Member Functions                                          */
/* ------------------------------------------------------------------------- */

Proctest::Proctest(int process_id)
{
    /* Record process ID */

    /* Initialize the read buffer */

    /* Read from the stat file */

    /* Read from the status file */

    /* Find the number of open file descriptors by counting /proc/[pid]/fd */

    /* Read Memory Map */
}


Proctest::~Proctest()
{
    // Free buffer memory & clear all vectors
}


std::string Proctest::getpid()
{
    // Dummy return statement
    return "";
}


std::string Proctest::getppid()
{
    // Dummy return statement
    return "";
}

std::string Proctest::geteuid()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getegid()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getruid()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getrgid()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getfsuid()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getfsgid()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getstate()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getthread_count()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getpriority()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getniceness()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getstime()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getutime()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getcstime()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getcutime()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getstartcode()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getendcode()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getesp()
{
    // Dummy return statement
    return "";
}

std::string Proctest::geteip()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getfiles()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getvoluntary_context_switches()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getnonvoluntary_context_switches()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getlast_cpu()
{
    // Dummy return statement
    return "";
}

std::string Proctest::getallowed_cpus()
{
    // Dummy return statement
    return "";
}

std::vector<std::string> Proctest::getmemory_map()
{
    // Dummy return statement
    return std::vector<std::string>();
}
